internal/testtypes/builders.go

```diff
@@ -14,164 +14,55 @@ See the License for the specific language governing permissions and
-			Group: fakeInfrastructureProviderGroupVersion.Group,
-			Names: apiextensionsv1.CustomResourceDefinitionNames{
-				Kind: "FakeInfrastructureClusterTemplate",
-			},
-		},
-	}
-	fakeControlPlaneTemplateCRD = &apiextensionsv1.CustomResourceDefinition{
-		TypeMeta: metav1.TypeMeta{
-			APIVersion: apiextensionsv1.SchemeGroupVersion.String(),
-			Kind:       "CustomResourceDefinition",
-		},
-		ObjectMeta: metav1.ObjectMeta{
-			Name: "fakecontrolplanetemplates.infrastructure.cluster.x-k8s.io",
-			Labels: map[string]string{
-				"cluster.x-k8s.io/v1alpha4": "v1alpha4",
-			},
-		},
-		Spec: apiextensionsv1.CustomResourceDefinitionSpec{
-			Group: fakeControlPlaneProviderGroupVersion.Group,
-			Names: apiextensionsv1.CustomResourceDefinitionNames{
-				Kind: "FakeControlPlaneTemplate",
-			},
-		},
-	}
-	fakeInfrastructureMachineTemplateCRD = &apiextensionsv1.CustomResourceDefinition{
-		TypeMeta: metav1.TypeMeta{
-			APIVersion: apiextensionsv1.SchemeGroupVersion.String(),
-			Kind:       "CustomResourceDefinition",
-		},
-		ObjectMeta: metav1.ObjectMeta{
-			Name: "fakeinfrastructuremachinetemplates.infrastructure.cluster.x-k8s.io",
-			Labels: map[string]string{
-				"cluster.x-k8s.io/v1alpha4": "v1alpha4",
-			},
-		},
-		Spec: apiextensionsv1.CustomResourceDefinitionSpec{
-			Group: fakeInfrastructureProviderGroupVersion.Group,
-			Names: apiextensionsv1.CustomResourceDefinitionNames{
-				Kind: "FakeInfrastructureMachineTemplate",
-			},
-		},
-	}
-	fakeBootstrapTemplateCRD = &apiextensionsv1.CustomResourceDefinition{
-		TypeMeta: metav1.TypeMeta{
-			APIVersion: apiextensionsv1.SchemeGroupVersion.String(),
-			Kind:       "CustomResourceDefinition",
-		},
-		ObjectMeta: metav1.ObjectMeta{
-			Name: "fakebootstraptemplates.infrastructure.cluster.x-k8s.io",
-			Labels: map[string]string{
-				"cluster.x-k8s.io/v1alpha4": "v1alpha4",
-			},
-		},
-		Spec: apiextensionsv1.CustomResourceDefinitionSpec{
-			Group: fakeBootstrapProviderGroupVersion.Group,
-			Names: apiextensionsv1.CustomResourceDefinitionNames{
-				Kind: "FakeBoostrapTemplate",
-			},
-		},
-	}
-)
-
-type fakeCluster struct {
+// ClusterBuilder holds the variables and objects required to build a clusterv1.Cluster.
+type ClusterBuilder struct {
 	namespace             string
 	name                  string
 	class                 clusterv1.ClusterClass
 	infrastructureCluster *unstructured.Unstructured
 	controlPlane          *unstructured.Unstructured
 }
 
-func newFakeCluster(namespace, name string) *fakeCluster {
-	return &fakeCluster{
+// NewClusterBuilder ...
+func NewClusterBuilder(namespace, name string) *ClusterBuilder {
+	return &ClusterBuilder{
 		namespace: namespace,
 		name:      name,
 	}
 }
 
-func (f *fakeCluster) WithInfrastructureCluster(t *unstructured.Unstructured) *fakeCluster {
+// WithInfrastructureCluster ...
+func (f *ClusterBuilder) WithInfrastructureCluster(t *unstructured.Unstructured) *ClusterBuilder {
```

Comment:
@killianmuldoon I'm mulling a little bit about this approach.

If we are passing the plain objects here this approach will suit only the case of building stand alone objects but not composition, thus requiring a rework in future.

I'm wondering if we should start accepting a builder here instead, like we are doing in https://github.com/kubernetes-sigs/cluster-api/blob/master/cmd/clusterctl/internal/test/fake_objects.go, so we can then support both generating stand alone objects and composition of entire cluster with extensive refactors

Reply:
```suggestion
func (f *ClusterBuilder) WithInfrastructureCluster(t *Builder) *ClusterBuilder {
```
Is the idea something like this? And have anInfrastructureClusterBuilder passed in here as a parameter?

Reply:
I think it's best to leave this for an iteration. I'd prefer to do it, even if it means rewrites, when we have a better idea of what the composition will look like in tests. Otherwise we run the risk of rewriting twice - once now, and once when we see what's really required here.